###       Lead Scoring Case Study 
####       Group Members :
           1. Balakrishna Gadiyar
           2. Deepak Kumar Padhan
           3. Bishnu Agrawal
 ======================================================================================================================         

#### Problem Statement  :

X Education sells online courses to industry professionals.The company markets its courses on several websites and search engine.Once these people land on the website, they might browse the courses or fill up a form for the course or watch some videos. When these people fill up a form providing their email address or phone number, they are classified to be a lead. 

Once these leads are acquired, employees from the sales team start making calls, writing emails, etc. Through this process, some of the leads get converted while most do not. The typical lead conversion rate at X education is around 30%.

X Education lead conversion rate is very poor.To make this process more efficient, the company wishes to identify the most potential leads, also known as ‘Hot Leads’.The company requires to build a model wherein it need to assign a lead score to each of the leads such that the customers with higher lead score have a higher conversion chance and the customers with lower lead score have a lower conversion chance. 



### Step 1: Importing and Merging Data


In [74]:
# Suppressing Warnings
import warnings
warnings.filterwarnings('ignore')

In [75]:
# Importing Pandas and NumPy
import pandas as pd, numpy as np

In [76]:
# Importing all datasets
lead_data = pd.read_csv("Leads.csv")


### Step 2: Inspecting the Dataframe

In [77]:
# Let's see the head of our master dataset

lead_data.head()

Prospect ID  Lead Number              Lead Origin  \
0  7927b2df-8bba-4d29-b9a2-b6e0beafe620       660737                      API   
1  2a272436-5132-4136-86fa-dcc88c88f482       660728                      API   
2  8cc8c611-a219-4f35-ad23-fdfd2656bd8a       660727  Landing Page Submission   
3  0cc2df48-7cf4-4e39-9de9-19797f9b38cc       660719  Landing Page Submission   
4  3256f628-e534-4826-9d63-4a8b88782852       660681  Landing Page Submission   

      Lead Source Do Not Email Do Not Call  Converted  TotalVisits  \
0      Olark Chat           No          No          0          0.0   
1  Organic Search           No          No          0          5.0   
2  Direct Traffic           No          No          1          2.0   
3  Direct Traffic           No          No          0          1.0   
4          Google           No          No          1          2.0   

   Total Time Spent on Website  Page Views Per Visit          ...           \
0                            0                   0.0          ...            
1                          674                   2.5          ...            
2                         1532                   2.0          ...            
3                          305                   1.0          ...            
4                         1428                   1.0          ...            

  Get updates on DM Content    Lead Profile    City  \
0                        No          Select  Select   
1                        No          Select  Select   
2                        No  Potential Lead  Mumbai   
3                        No          Select  Mumbai   
4                        No          Select  Mumbai   

  Asymmetrique Activity Index Asymmetrique Profile Index  \
0                   02.Medium                  02.Medium   
1                   02.Medium                  02.Medium   
2                   02.Medium                    01.High   
3                   02.Medium                    01.High   
4                   02.Medium                    01.High   

  Asymmetrique Activity Score Asymmetrique Profile Score  \
0                        15.0                       15.0   
1                        15.0                       15.0   
2                        14.0                       20.0   
3                        13.0                       17.0   
4                        15.0                       18.0   

  I agree to pay the amount through cheque  \
0                                       No   
1                                       No   
2                                       No   
3                                       No   
4                                       No   

  A free copy of Mastering The Interview Last Notable Activity  
0                                     No              Modified  
1                                     No          Email Opened  
2                                    Yes          Email Opened  
3                                     No              Modified  
4                                     No              Modified  

[5 rows x 37 columns]

In [78]:
# Check Sanity of data .
lead_data.shape

(9240, 37)

In [79]:
#Descirbe the dataframe
lead_data.describe()

Lead Number    Converted  TotalVisits  Total Time Spent on Website  \
count    9240.000000  9240.000000  9103.000000                  9240.000000   
mean   617188.435606     0.385390     3.445238                   487.698268   
std     23405.995698     0.486714     4.854853                   548.021466   
min    579533.000000     0.000000     0.000000                     0.000000   
25%    596484.500000     0.000000     1.000000                    12.000000   
50%    615479.000000     0.000000     3.000000                   248.000000   
75%    637387.250000     1.000000     5.000000                   936.000000   
max    660737.000000     1.000000   251.000000                  2272.000000   

       Page Views Per Visit  Asymmetrique Activity Score  \
count           9103.000000                  5022.000000   
mean               2.362820                    14.306252   
std                2.161418                     1.386694   
min                0.000000                     7.000000   
25%                1.000000                    14.000000   
50%                2.000000                    14.000000   
75%                3.000000                    15.000000   
max               55.000000                    18.000000   

       Asymmetrique Profile Score  
count                 5022.000000  
mean                    16.344883  
std                      1.811395  
min                     11.000000  
25%                     15.000000  
50%                     16.000000  
75%                     18.000000  
max                     20.000000

In [80]:
#Review the info from dataframe- understand column types
lead_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9240 entries, 0 to 9239
Data columns (total 37 columns):
Prospect ID                                      9240 non-null object
Lead Number                                      9240 non-null int64
Lead Origin                                      9240 non-null object
Lead Source                                      9204 non-null object
Do Not Email                                     9240 non-null object
Do Not Call                                      9240 non-null object
Converted                                        9240 non-null int64
TotalVisits                                      9103 non-null float64
Total Time Spent on Website                      9240 non-null int64
Page Views Per Visit                             9103 non-null float64
Last Activity                                    9137 non-null object
Country                                          6779 non-null object
Specialization                                   7802 

In [81]:
#Since select are also Null values, lets replace all select with NANs
lead_data.replace('Select', np.nan,inplace = True)

In [82]:
# Check the columns having Null value ..
lead_data.isnull().values.any()

True

In [83]:
lead_data.head()

Prospect ID  Lead Number              Lead Origin  \
0  7927b2df-8bba-4d29-b9a2-b6e0beafe620       660737                      API   
1  2a272436-5132-4136-86fa-dcc88c88f482       660728                      API   
2  8cc8c611-a219-4f35-ad23-fdfd2656bd8a       660727  Landing Page Submission   
3  0cc2df48-7cf4-4e39-9de9-19797f9b38cc       660719  Landing Page Submission   
4  3256f628-e534-4826-9d63-4a8b88782852       660681  Landing Page Submission   

      Lead Source Do Not Email Do Not Call  Converted  TotalVisits  \
0      Olark Chat           No          No          0          0.0   
1  Organic Search           No          No          0          5.0   
2  Direct Traffic           No          No          1          2.0   
3  Direct Traffic           No          No          0          1.0   
4          Google           No          No          1          2.0   

   Total Time Spent on Website  Page Views Per Visit          ...           \
0                            0                   0.0          ...            
1                          674                   2.5          ...            
2                         1532                   2.0          ...            
3                          305                   1.0          ...            
4                         1428                   1.0          ...            

  Get updates on DM Content    Lead Profile    City  \
0                        No             NaN     NaN   
1                        No             NaN     NaN   
2                        No  Potential Lead  Mumbai   
3                        No             NaN  Mumbai   
4                        No             NaN  Mumbai   

  Asymmetrique Activity Index Asymmetrique Profile Index  \
0                   02.Medium                  02.Medium   
1                   02.Medium                  02.Medium   
2                   02.Medium                    01.High   
3                   02.Medium                    01.High   
4                   02.Medium                    01.High   

  Asymmetrique Activity Score Asymmetrique Profile Score  \
0                        15.0                       15.0   
1                        15.0                       15.0   
2                        14.0                       20.0   
3                        13.0                       17.0   
4                        15.0                       18.0   

  I agree to pay the amount through cheque  \
0                                       No   
1                                       No   
2                                       No   
3                                       No   
4                                       No   

  A free copy of Mastering The Interview Last Notable Activity  
0                                     No              Modified  
1                                     No          Email Opened  
2                                    Yes          Email Opened  
3                                     No              Modified  
4                                     No              Modified  

[5 rows x 37 columns]

In [84]:
lead_data.isnull().sum()

Prospect ID                                         0
Lead Number                                         0
Lead Origin                                         0
Lead Source                                        36
Do Not Email                                        0
Do Not Call                                         0
Converted                                           0
TotalVisits                                       137
Total Time Spent on Website                         0
Page Views Per Visit                              137
Last Activity                                     103
Country                                          2461
Specialization                                   3380
How did you hear about X Education               7250
What is your current occupation                  2690
What matters most to you in choosing a course    2709
Search                                              0
Magazine                                            0
Newspaper Article           

In [85]:
#Percentage of nulls in the data set
lead_data.isnull().sum()*100/lead_data.shape[0]

Prospect ID                                       0.000000
Lead Number                                       0.000000
Lead Origin                                       0.000000
Lead Source                                       0.389610
Do Not Email                                      0.000000
Do Not Call                                       0.000000
Converted                                         0.000000
TotalVisits                                       1.482684
Total Time Spent on Website                       0.000000
Page Views Per Visit                              1.482684
Last Activity                                     1.114719
Country                                          26.634199
Specialization                                   36.580087
How did you hear about X Education               78.463203
What is your current occupation                  29.112554
What matters most to you in choosing a course    29.318182
Search                                            0.0000

### Step 3: Data Preparation

#### Drop not useful Columns

In [86]:
df_lead=lead_data

df_lead.columns

Index(['Prospect ID', 'Lead Number', 'Lead Origin', 'Lead Source',
       'Do Not Email', 'Do Not Call', 'Converted', 'TotalVisits',
       'Total Time Spent on Website', 'Page Views Per Visit', 'Last Activity',
       'Country', 'Specialization', 'How did you hear about X Education',
       'What is your current occupation',
       'What matters most to you in choosing a course', 'Search', 'Magazine',
       'Newspaper Article', 'X Education Forums', 'Newspaper',
       'Digital Advertisement', 'Through Recommendations',
       'Receive More Updates About Our Courses', 'Tags', 'Lead Quality',
       'Update me on Supply Chain Content', 'Get updates on DM Content',
       'Lead Profile', 'City', 'Asymmetrique Activity Index',
       'Asymmetrique Profile Index', 'Asymmetrique Activity Score',
       'Asymmetrique Profile Score',
       'I agree to pay the amount through cheque',
       'A free copy of Mastering The Interview', 'Last Notable Activity'],
      dtype='object')

In [87]:
#### Drop Column having same value as 'No' ; as this is not be useful in deciding whether a lead will be converted or not.
#### 'Magazine','Receive More Updates About Our Courses','Update me on Supply Chain Content', 'Get updates on DM Content', 'I agree to pay the amount through cheque'

dropcols=['Prospect ID','Lead Number','Magazine','Receive More Updates About Our Courses','Update me on Supply Chain Content', 'Get updates on DM Content', 'I agree to pay the amount through cheque']
df_lead.drop(dropcols, axis=1, inplace=True)
df_lead.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9240 entries, 0 to 9239
Data columns (total 30 columns):
Lead Origin                                      9240 non-null object
Lead Source                                      9204 non-null object
Do Not Email                                     9240 non-null object
Do Not Call                                      9240 non-null object
Converted                                        9240 non-null int64
TotalVisits                                      9103 non-null float64
Total Time Spent on Website                      9240 non-null int64
Page Views Per Visit                             9103 non-null float64
Last Activity                                    9137 non-null object
Country                                          6779 non-null object
Specialization                                   5860 non-null object
How did you hear about X Education               1990 non-null object
What is your current occupation                  6550

# Dropping columns with more than 40% of missing values.
##### How did you hear about X Education               78.463203
##### Lead Quality                                     51.590909
##### Lead Profile                                     74.188312
##### City                                             39.707792
##### Asymmetrique Activity Index                      45.649351
##### Asymmetrique Profile Index                       45.649351
##### Asymmetrique Activity Score                      45.649351
##### Asymmetrique Profile Score                       45.649351


In [88]:
dropcols=['Asymmetrique Activity Index','Asymmetrique Profile Index','Asymmetrique Activity Score', 'Asymmetrique Profile Score', 'How did you hear about X Education','Lead Quality','Lead Profile','City']
df_lead.drop(dropcols, axis=1, inplace=True)
df_lead.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9240 entries, 0 to 9239
Data columns (total 22 columns):
Lead Origin                                      9240 non-null object
Lead Source                                      9204 non-null object
Do Not Email                                     9240 non-null object
Do Not Call                                      9240 non-null object
Converted                                        9240 non-null int64
TotalVisits                                      9103 non-null float64
Total Time Spent on Website                      9240 non-null int64
Page Views Per Visit                             9103 non-null float64
Last Activity                                    9137 non-null object
Country                                          6779 non-null object
Specialization                                   5860 non-null object
What is your current occupation                  6550 non-null object
What matters most to you in choosing a course    6531

#### Treat the 'Blank'  with Null

In [89]:
# List of Columns with 'Blank' and 'Select'
df_lead.head()

Lead Origin     Lead Source Do Not Email Do Not Call  \
0                      API      Olark Chat           No          No   
1                      API  Organic Search           No          No   
2  Landing Page Submission  Direct Traffic           No          No   
3  Landing Page Submission  Direct Traffic           No          No   
4  Landing Page Submission          Google           No          No   

   Converted  TotalVisits  Total Time Spent on Website  Page Views Per Visit  \
0          0          0.0                            0                   0.0   
1          0          5.0                          674                   2.5   
2          1          2.0                         1532                   2.0   
3          0          1.0                          305                   1.0   
4          1          2.0                         1428                   1.0   

             Last Activity Country          ...           \
0  Page Visited on Website     NaN          ...            
1             Email Opened   India          ...            
2             Email Opened   India          ...            
3              Unreachable   India          ...            
4        Converted to Lead   India          ...            

  What matters most to you in choosing a course Search Newspaper Article  \
0                       Better Career Prospects     No                No   
1                       Better Career Prospects     No                No   
2                       Better Career Prospects     No                No   
3                       Better Career Prospects     No                No   
4                       Better Career Prospects     No                No   

  X Education Forums Newspaper Digital Advertisement Through Recommendations  \
0                 No        No                    No                      No   
1                 No        No                    No                      No   
2                 No        No                    No                      No   
3                 No        No                    No                      No   
4                 No        No                    No                      No   

                                  Tags A free copy of Mastering The Interview  \
0          Interested in other courses                                     No   
1                              Ringing                                     No   
2  Will revert after reading the email                                    Yes   
3                              Ringing                                     No   
4  Will revert after reading the email                                     No   

  Last Notable Activity  
0              Modified  
1          Email Opened  
2          Email Opened  
3              Modified  
4              Modified  

[5 rows x 22 columns]

In [90]:
df_lead.drop(['Tags','TotalVisits','Page Views Per Visit'], axis=1, inplace=True)

In [91]:
df_lead.replace('Select', np.nan,inplace = True)
df_lead['Country'].replace(np.nan, 'Unknown',inplace = True)
#df_lead['City'].replace([np.nan,'Select'], 'Other Cities',inplace = True)
df_lead['Specialization'].replace([np.nan,'Select'], 'Not Specified',inplace = True)
df_lead['Lead Source'].replace([np.nan,'Select'], 'Unknown',inplace = True)
df_lead['Last Activity'].replace([np.nan,'Select'], 'Unknown',inplace = True)
#df_lead['How did you hear about X Education'].replace([np.nan,'Select'], 'Not Specified',inplace = True)
#df_lead['TotalVisits'].replace([np.nan,'Select'], '0',inplace = True)
#df_lead['Page Views Per Visit'].replace([np.nan,'Select'], '0',inplace = True)
df_lead['What is your current occupation'].replace([np.nan,'Select'], 'Not Disclosed',inplace = True)
df_lead['What matters most to you in choosing a course'].replace([np.nan,'Select'], 'Other',inplace = True)
#df_lead['Lead Quality'].replace([np.nan,'Select'], 'Not Sure',inplace = True)
#df_lead['Lead Profile'].replace([np.nan,'Select'], 'Other Leads',inplace = True)
df_lead['Country'].replace([np.nan,'Select'], 'Other Cities',inplace = True)
#df_lead['Asymmetrique Activity Score'].replace([np.nan,'Select'], round(df_lead["Asymmetrique Activity Score"].mean(),0),inplace = True)
#df_lead['Asymmetrique Profile Score'].replace([np.nan,'Select'], round(df_lead["Asymmetrique Profile Score"].mean(),0),inplace = True)



In [92]:
df_lead.isnull().sum()

Lead Origin                                      0
Lead Source                                      0
Do Not Email                                     0
Do Not Call                                      0
Converted                                        0
Total Time Spent on Website                      0
Last Activity                                    0
Country                                          0
Specialization                                   0
What is your current occupation                  0
What matters most to you in choosing a course    0
Search                                           0
Newspaper Article                                0
X Education Forums                               0
Newspaper                                        0
Digital Advertisement                            0
Through Recommendations                          0
A free copy of Mastering The Interview           0
Last Notable Activity                            0
dtype: int64

#### Converting  binary variables (Yes/No) to 0/1

In [93]:
# List of variables to map

varlist =  ['Do Not Email', 'Do Not Call', 'Search','Newspaper Article', 'X Education Forums','Newspaper','Digital Advertisement','Through Recommendations','A free copy of Mastering The Interview']
# Defining the map function
def binary_map(x):
    return x.map({'yes': 1, "no": 0,'Yes': 1, "No": 0})

# Applying the function to the housing list
df_lead[varlist] = df_lead[varlist].apply(binary_map)


In [94]:
df_lead

Lead Origin     Lead Source  Do Not Email  Do Not Call  \
0                         API      Olark Chat             0            0   
1                         API  Organic Search             0            0   
2     Landing Page Submission  Direct Traffic             0            0   
3     Landing Page Submission  Direct Traffic             0            0   
4     Landing Page Submission          Google             0            0   
5                         API      Olark Chat             0            0   
6     Landing Page Submission          Google             0            0   
7                         API      Olark Chat             0            0   
8     Landing Page Submission  Direct Traffic             0            0   
9                         API          Google             0            0   
10    Landing Page Submission  Organic Search             0            0   
11    Landing Page Submission  Direct Traffic             0            0   
12                        API  Organic Search             0            0   
13    Landing Page Submission  Organic Search             0            0   
14    Landing Page Submission  Direct Traffic             1            0   
15                        API  Organic Search             0            0   
16                        API      Olark Chat             0            0   
17                        API  Referral Sites             0            0   
18    Landing Page Submission          Google             0            0   
19                        API  Organic Search             0            0   
20    Landing Page Submission          Google             0            0   
21                        API          Google             0            0   
22    Landing Page Submission          Google             0            0   
23    Landing Page Submission          Google             0            0   
24                        API          Google             0            0   
25    Landing Page Submission          Google             0            0   
26    Landing Page Submission  Organic Search             0            0   
27    Landing Page Submission          Google             0            0   
28    Landing Page Submission  Direct Traffic             0            0   
29                        API          Google             0            0   
...                       ...             ...           ...          ...   
9210  Landing Page Submission  Direct Traffic             0            0   
9211  Landing Page Submission  Direct Traffic             0            0   
9212  Landing Page Submission          Google             1            0   
9213  Landing Page Submission  Direct Traffic             1            0   
9214                      API  Organic Search             0            0   
9215  Landing Page Submission  Organic Search             0            0   
9216  Landing Page Submission  Direct Traffic             1            0   
9217                      API      Olark Chat             0            0   
9218  Landing Page Submission          Google             1            0   
9219  Landing Page Submission  Direct Traffic             0            0   
9220  Landing Page Submission  Direct Traffic             0            0   
9221  Landing Page Submission          Google             0            0   
9222                      API          Google             0            0   
9223                      API  Organic Search             0            0   
9224  Landing Page Submission          Google             0            0   
9225  Landing Page Submission  Direct Traffic             1            0   
9226                      API      Olark Chat             0            0   
9227  Landing Page Submission          Google             0            0   
9228  Landing Page Submission          Google             0            0   
9229  Landing Page Submission  Organic Search             0            0   
9230  Landing Page Submission          Google             0       

In [95]:
# Creating dummy variable for variable risk and dropping the first one
LeadOrigin = pd.get_dummies(df_lead['Lead Origin'],drop_first=True)
#Adding the results to the master dataframe
df_lead = pd.concat([df_lead,LeadOrigin],axis=1)
# Dropping risk as we have created the dummies for it
df_lead.drop(['Lead Origin'],axis=1,inplace=True)

#Repaeting the steps for Lead Source
LeadSource = pd.get_dummies(df_lead['Lead Source'],drop_first=True)
df_lead = pd.concat([df_lead,LeadSource],axis=1)
df_lead.drop(['Lead Source'],axis=1,inplace=True)

#Repaeting the steps for Last Activity
LastActivity = pd.get_dummies(df_lead['Last Activity'],drop_first=True)
df_lead = pd.concat([df_lead,LastActivity],axis=1)
df_lead.drop(['Last Activity'],axis=1,inplace=True)

#Repaeting the steps for Country

Country = pd.get_dummies(df_lead['Country'],drop_first=True)
df_lead = pd.concat([df_lead,Country],axis=1)
df_lead.drop(['Country'],axis=1,inplace=True)


#Repaeting the steps for Specialization
Specialization = pd.get_dummies(df_lead['Specialization'],drop_first=True)
df_lead = pd.concat([df_lead,Specialization],axis=1)
df_lead.drop(['Specialization'],axis=1,inplace=True)


#Repaeting the steps for How did you hear about X Education
#HowdidyouhearaboutXEducation = pd.get_dummies(df_lead['How did you hear about X Education'],drop_first=True)
#df_lead = pd.concat([df_lead,HowdidyouhearaboutXEducation],axis=1)
#df_lead.drop(['How did you hear about X Education'],axis=1,inplace=True)



#Repaeting the steps for What is your current occupation
Whatisyourcurrentoccupation = pd.get_dummies(df_lead['What is your current occupation'],drop_first=True)
df_lead = pd.concat([df_lead,Whatisyourcurrentoccupation],axis=1)
df_lead.drop(['What is your current occupation'],axis=1,inplace=True)


#Repaeting the steps for What matters most to you in choosing a course
Whatmattersmosttoyouinchoosingacourse = pd.get_dummies(df_lead['What matters most to you in choosing a course'],drop_first=True)
df_lead = pd.concat([df_lead,Whatmattersmosttoyouinchoosingacourse],axis=1)
df_lead.drop(['What matters most to you in choosing a course'],axis=1,inplace=True)


#Repaeting the steps for Lead Profile
#LeadProfile = pd.get_dummies(df_lead['Lead Profile'],drop_first=True)
#df_lead = pd.concat([df_lead,LeadProfile],axis=1)
#df_lead.drop(['Lead Profile'],axis=1,inplace=True)

#Repaeting the steps for City
#City = pd.get_dummies(df_lead['City'],drop_first=True)
#df_lead = pd.concat([df_lead,City],axis=1)
#df_lead.drop(['City'],axis=1,inplace=True)


#Repaeting the steps for Last Notable Activity
LastNotableActivity = pd.get_dummies(df_lead['Last Notable Activity'],drop_first=True)
df_lead = pd.concat([df_lead,LastNotableActivity],axis=1)
df_lead.drop(['Last Notable Activity'],axis=1,inplace=True)




df_lead.head()

Do Not Email  Do Not Call  Converted  Total Time Spent on Website  Search  \
0             0            0          0                            0       0   
1             0            0          0                          674       0   
2             0            0          1                         1532       0   
3             0            0          0                          305       0   
4             0            0          1                         1428       0   

   Newspaper Article  X Education Forums  Newspaper  Digital Advertisement  \
0                  0                   0          0                      0   
1                  0                   0          0                      0   
2                  0                   0          0                      0   
3                  0                   0          0                      0   
4                  0                   0          0                      0   

   Through Recommendations              ...               \
0                        0              ...                
1                        0              ...                
2                        0              ...                
3                        0              ...                
4                        0              ...                

   Form Submitted on Website  Had a Phone Conversation  Modified  \
0                          0                         0         1   
1                          0                         0         0   
2                          0                         0         0   
3                          0                         0         1   
4                          0                         0         1   

   Olark Chat Conversation  Page Visited on Website  Resubscribed to emails  \
0                        0                        0                       0   
1                        0                        0                       0   
2                        0                        0                       0   
3                        0                        0                       0   
4                        0                        0                       0   

   SMS Sent  Unreachable  Unsubscribed  View in browser link Clicked  
0         0            0             0                             0  
1         0            0             0                             0  
2         0            0             0                             0  
3         0            0             0                             0  
4         0            0             0                             0  

[5 rows x 132 columns]

In [96]:
df_lead.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9240 entries, 0 to 9239
Columns: 132 entries, Do Not Email to View in browser link Clicked
dtypes: int64(11), uint8(121)
memory usage: 1.8 MB


### Normalizing data

In [97]:
#defining a normalisation function 
def normalize (x): 
    return ( (x-np.min(x))/ (max(x) - min(x)))
                                            
                                              
# applying normalize ( ) to all columns 
df_lead = df_lead.apply(normalize) 

In [98]:
df_lead

Do Not Email  Do Not Call  Converted  Total Time Spent on Website  \
0              0.0          0.0        0.0                     0.000000   
1              0.0          0.0        0.0                     0.296655   
2              0.0          0.0        1.0                     0.674296   
3              0.0          0.0        0.0                     0.134243   
4              0.0          0.0        1.0                     0.628521   
5              0.0          0.0        0.0                     0.000000   
6              0.0          0.0        1.0                     0.721831   
7              0.0          0.0        0.0                     0.000000   
8              0.0          0.0        0.0                     0.031250   
9              0.0          0.0        0.0                     0.025528   
10             0.0          0.0        1.0                     0.594630   
11             0.0          0.0        1.0                     0.591109   
12             0.0          0.0        1.0                     0.676937   
13             0.0          0.0        0.0                     0.074824   
14             1.0          0.0        0.0                     0.211708   
15             0.0          0.0        1.0                     0.445423   
16             0.0          0.0        0.0                     0.000000   
17             0.0          0.0        0.0                     0.428257   
18             0.0          0.0        1.0                     0.742958   
19             0.0          0.0        0.0                     0.043134   
20             0.0          0.0        0.0                     0.102553   
21             0.0          0.0        0.0                     0.165933   
22             0.0          0.0        1.0                     0.445863   
23             0.0          0.0        0.0                     0.339349   
24             0.0          0.0        1.0                     0.500440   
25             0.0          0.0        1.0                     0.470070   
26             0.0          0.0        1.0                     0.440141   
27             0.0          0.0        1.0                     0.578785   
28             0.0          0.0        0.0                     0.080106   
29             0.0          0.0        1.0                     0.034331   
...            ...          ...        ...                          ...   
9210           0.0          0.0        1.0                     0.408011   
9211           0.0          0.0        1.0                     0.489437   
9212           1.0          0.0        0.0                     0.034331   
9213           1.0          0.0        0.0                     0.102993   
9214           0.0          0.0        1.0                     0.387764   
9215           0.0          0.0        0.0                     0.174736   
9216           1.0          0.0        0.0                     0.738996   
9217           0.0          0.0        0.0                     0.000000   
9218           1.0          0.0        0.0                     0.065581   
9219           0.0          0.0        1.0                     0.611356   
9220           0.0          0.0        0.0                     0.008803   
9221           0.0          0.0        0.0                     0.592870   
9222           0.0          0.0        0.0                     0.100352   
9223           0.0          0.0        0.0                     0.062500   
9224           0.0          0.0        0.0                     0.200264   
9225           1.0          0.0        0.0                     0.032570   
9226           0.0          0.0        0.0                     0.000000   
9227           0.0          0.0        1.0                     0.564701   
9228           0.0          0.0        1.0                     0.855634   
9229           0.0          0.0        1.0                     0.539613   
9230           0.0          0.0        0.0                     0.382923   
9231           0.0          0.0